### ISSUE TITLE
Possible coordinate system mismatch with Vessel::facing and ManeuverNode::burn_vector

### ISSUE THREAD

**Issue by kbleeke** [untoldwind/KontrolSystem2#54]
I'm (sometimes) having trouble with `ship_is_facing` (called by exec_next_node). 
The vessel is clearly pointing the corect direction according to the navball. 
However, the vectors printed out by the script (facing, and burn_vector) don't seem to be pointing in the same direction, so the desired angle of <5° is never reached.
 Sometimes this seems to work, sometimes it doesn't.

Using their `global_` siblings  and manually transform them into e.g. the control_frame seems to give reasonable results, so that leads me to believe that `facing` and `burn_vector` (sometimes?) don' use the same reference frame?

**Comment by kbleeke** [comment]
Example in Jool Orbit: Vessel clearly pointing in the correct direction. Vectors are not.

![image](https://user-images.githubusercontent.com/39027073/227528695-de4ee5ad-2ba4-436a-adc5-605fc97f6d37.png)

**Comment by untoldwind** [comment]
This probably needs some deeper investigation. Internally the maneuver node uses the `orbit.reference_frame` whereas the `vessel.facing` uses the `vessel.main_body.celestial_frame`.
I had the impression, that `orbit.reference_frame` is always the same as `orbit.reference_body.celestial_frame`, maybe that is not the case here for some reason.
(... or is `orbit.reference_body != vessel.main_body` in this case?)

It would probably best to patch the `exec_node` script to use the `global_` vectors instead.

**Comment by kbleeke** [comment]
Hyperbolic Dres entry, right at the edge 

![image](https://user-images.githubusercontent.com/39027073/227623640-29a77a6c-3aab-432e-9262-211fb10c73af.png)

This is the readout of the various variants of the vectors. It seems facing steps out of line here, actually. And burn vector in celestial is slightly off?

**Comment by kbleeke** [comment]
It get a bit weirder as the celestial frame seems to change (on Dres) somewhere between 24km and 25km. When flying below that altitude global_facing in orbit.reference_frame and global_facing in celestial_frame are no longer equal. Looks like the same behavior as I observed in #4

**Comment by untoldwind** [comment]
That is quite the find. I still don't understand why line 2 and 4 differ:
If 
```
node.global_burn_vector.to_local(vessel.orbit.reference_frame) == vessel.global_facing.vector.to_local(vessel.orbit.reference_frame)
```
then it should be
```
node.global_burn_vector.to_local(vessel.celestial_frame) ==
vessel.global_facing.vector.to_local(vessel.celestial_frame)
```
(unless some there is some wired timing issue.)
But line 5 and 6 are the important ones:
Obviously `orbit.reference_frame` is the same (or mostly the same) as `vessel.celestial_frame` and not `vessel.main_body.celestial_frame` ... it might even depend on if you are using `vessel.orbit` vs. `body.orbit`.
In short it is not very relyable.

I will change `node.burn_vector` (i.e. the not-global one) to use `orbit.reference_body.celestial_frame` instead. That should fix the difference.

**Comment by untoldwind** [comment]
Have to some testing on that change, but it seems the most consistent way is to use the celestial frame of the main body whenever possible.

**Comment by kbleeke** [comment]
Ah this was my bad, I used vessel.celestial_frame instead of vessel.main_body.celestial_frame.

This is the same location as before but with vessel.main_body.celestial_frame, these actually seem to be the same.

![image](https://user-images.githubusercontent.com/39027073/227652060-1bb4bb25-1ce6-421d-8ecc-0a18d20c6196.png)

**Comment by kbleeke** [comment]
As far as I can tell, It's actually `vessel.facing` that is using the body frame not the celestial. https://github.com/untoldwind/KontrolSystem2/blob/master/KSP2Runtime/KSPVessel/KSPVesselModule.Vessel.cs#L155

I assume, this is desired in some locations? Maybe there is nothing to do but explicitely transform facing in the celestial frame in the script.

**Comment by untoldwind** [comment]
Ah, that is a little trick I came up with, I take the identity rotation in the body-frame of the vessel and transform it to a rotation in the celestial frame of the body, so I get the rotation of the vessel relative to the main body. The rest of the code is dealing with the awkward orientation of the vessel coordinates that have seemingly been taken over from KSP1.

**Comment by kbleeke** [comment]
Ok this seems to work now, thanks. I'm slowly wrapping my head around all the different coordinate systems. I tested with my Dres prope and a short trip from Kerbin to Mun Orbit.

(shoutout to my Dres probe that can now finally crash into Dres because it seems to be experiencing drag from the non-existant atomosphere)

**Comment by untoldwind** [comment]
You are not alone in this coordinate system nightmare ;)
On the other hand KSP1 pretty much had the same issues, it was just way less obvious and one sometimes had to some deep digging to figure out what reference frame that Vector3d was actually based on.
